File: bake.py
#!/usr/bin/python3

import argparse
import dataclasses
import json
import logging
import pathlib
import sys

__version__ = "0.1.0"
logger = logging.getLogger()


@dataclasses.dataclass
class Platform:
    os: str
    arch: str


def main(argv):
    args = _parse_argv(argv[1:])
    _add_handlers(logger, [], level_stderr=args.log_level)
    logger.debug(dict(args=args))
    run(args)


def run(args):
    zstd = "compression=zstd,force-compression=true"
    spec = dict(group=dict(), target=dict())
    for platform in (
        Platform(os="linux", arch="amd64"),
        Platform(os="linux", arch="arm64"),
    ):
        test_ks = []
        prod_ks = []
        for target, image_name in (
            ("test_client", "/client/test"),
            ("prod_envoy", "/envoy"),
            ("prod_nginx", "/nginx"),
            ("prod_postgres", "/postgres"),
            ("prod_postgres_migration", "/postgres_migration"),
            ("prod_api_v1", "/api_v1"),
            ("tests_e2e", "/tests_e2e"),
        ):
            k = f"{target}-{platform.os}-{platform.arch}"
            v = {
                "dockerfile": f"Dockerfile",
                "target": target,
                "output": [
                    f"type=docker,{zstd}",
                ],
                "tags": [
                    f"{args.base}{image_name}:h-{args.sha}-{platform.os}-{platform.arch}",
                ],
                "platforms": [
                    f"{platform.os}/{platform.arch}",
                ],
                "args": dict(arch=platform.arch),
                "cache-from": [
                    f"type=registry,ref={args.base}{image_name}/cache:{args.ref_b64}-{platform.os}-{platform.arch}",
                    f"type=registry,ref={args.base}{image_name}/cache:latest-{platform.os}-{platform.arch}",
                    f"type=registry,ref={args.base}{image_name}:{args.ref_b64}-{platform.os}-{platform.arch}",
                    f"type=registry,ref={args.base}{image_name}:latest-{platform.os}-{platform.arch}",
                ],
                "cache-to": [
                    f"type=registry,ref={args.base}{image_name}/cache:{args.ref_b64}-{platform.os}-{platform.arch},mode=max,{zstd},ignore-error={args.ignore_cache_error}",
                    f"type=registry,ref={args.base}{image_name}/cache:latest-{platform.os}-{platform.arch},mode=max,{zstd},ignore-error={args.ignore_cache_error}",
                ],
            }
            spec["target"][k] = v
            if k.startswith("prod_"):
                prod_ks.append(k)
            else:
                test_ks.append(k)
        spec["group"][f"{platform.os}-{platform.arch}-prod"] = dict(targets=prod_ks)
        spec["group"][f"{platform.os}-{platform.arch}-test"] = dict(targets=test_ks)
    json.dump(
        spec,
        sys.stdout,
        indent=2,
        sort_keys=True,
        ensure_ascii=False,
    )


def _parse_argv(argv):
    logger.debug(dict(argv=argv))
    doc = f"""
    {__file__}
    """

    parser = argparse.ArgumentParser(
        doc, formatter_class=argparse.ArgumentDefaultsHelpFormatter
    )
    parser.add_argument(
        "--version", action="version", version=f"%(prog)s {__version__}"
    )
    parser.add_argument(
        "--log_level",
        default="warning",
        type=lambda x: getattr(logging, x.upper()),
        help="Set log level.",
    )
    parser.add_argument("--sha", required=True)
    parser.add_argument("--ref_b64", required=True)
    parser.add_argument("--base", default="ghcr.io/kshramt/evidence_based_scheduling")
    parser.add_argument("--ignore_cache_error", default="false")
    args = parser.parse_args(argv)
    logger.debug(dict(args=args))
    return args


def _add_handlers(logger, paths, level_stderr=logging.INFO, level_path=logging.DEBUG):
    fmt = logging.Formatter(
        "%(levelname)s\t%(process)d\t%(asctime)s\t%(name)s\t%(funcName)s\t%(lineno)d\t%(message)s"
    )
    import time

    fmt.converter = time.gmtime
    fmt.default_time_format = "%Y-%m-%dT%H:%M:%S"
    fmt.default_msec_format = "%s.%03dZ"

    hdl = logging.StreamHandler(sys.stderr)
    hdl.setFormatter(fmt)
    hdl.setLevel(level_stderr)
    logger.setLevel(logging.DEBUG)
    logger.addHandler(hdl)

    for path in paths:
        pathlib.Path(path).parent.mkdir(exist_ok=True, parents=True)
        hdl_ = logging.FileHandler(path)
        hdl_.setFormatter(fmt)
        hdl_.setLevel(level_path)
        logger.addHandler(hdl_)

    logger.info(dict(log_files=paths))
    return logger


if __name__ == "__main__":
    main(sys.argv)

File: bake.ts
#!/usr/bin/env -S deno run --allow-all

const usage_and_exit = () => {
  console.error(
    "Usage: bake.ts <sha> <ref_b64> [<base>] [<ignore_cache_error>]"
  );
  Deno.exit(1);
};

const run = (
  sha: string,
  ref_b64: string,
  base: string,
  ignore_cache_error: boolean
) => {
  const zstd = "compression=zstd,force-compression=true";
  const spec: {
    group: Record<string, { targets: string[] }>;
    target: Record<
      string,
      {
        "cache-from": string[];
        "cache-to": string[];
        args: Record<string, string>;
        dockerfile: string;
        output: string[];
        platforms: string[];
        tags: string[];
        target: string;
      }
    >;
  } = { group: {}, target: {} };
  for (const platform of [
    { os: "linux", arch: "amd64" },
    { os: "linux", arch: "arm64" },
  ]) {
    const test_ks: string[] = [];
    const prod_ks: string[] = [];
    for (const [target, image_name] of [
      ["prod_api_v2", "/api_v2"],
      ["prod_envoy", "/envoy"],
      ["prod_nginx", "/nginx"],
      ["prod_postgres_migration", "/postgres_migration"],
      ["prod_postgres", "/postgres"],
      ["test_client", "/client/test"],
      ["tests_e2e", "/tests_e2e"],
    ] as const) {
      const k = `${target}-${platform.os}-${platform.arch}`;
      const v = {
        "cache-from": [
          `type=registry,ref=${base}${image_name}/cache:${ref_b64}-${platform.os}-${platform.arch}`,
          `type=registry,ref=${base}${image_name}/cache:latest-${platform.os}-${platform.arch}`,
          `type=registry,ref=${base}${image_name}:${ref_b64}-${platform.os}-${platform.arch}`,
          `type=registry,ref=${base}${image_name}:latest-${platform.os}-${platform.arch}`,
        ],
        "cache-to": [
          `type=registry,ref=${base}${image_name}/cache:${ref_b64}-${platform.os}-${platform.arch},mode=max,${zstd},ignore-error=${ignore_cache_error}`,
          `type=registry,ref=${base}${image_name}/cache:latest-${platform.os}-${platform.arch},mode=max,${zstd},ignore-error=${ignore_cache_error}`,
        ],
        dockerfile: "Dockerfile",
        output: [`type=docker,${zstd}`],
        platforms: [`${platform.os}/${platform.arch}`],
        tags: [`${base}${image_name}:h-${sha}-${platform.os}-${platform.arch}`],
        target,
        args: { arch: platform.arch },
      };
      spec.target[k] = v;
      if (k.startsWith("prod_")) {
        prod_ks.push(k);
      } else {
        test_ks.push(k);
      }
    }
    spec.group[`${platform.os}-${platform.arch}-prod`] = { targets: prod_ks };
    spec.group[`${platform.os}-${platform.arch}-test`] = { targets: test_ks };
  }
  console.log(JSON.stringify(spec, null, 2));
};

const main = (args: string[]) => {
  if (args.length < 2 || 4 < args.length) {
    usage_and_exit();
  }
  const sha = args[0];
  const ref_b64 = args[1];
  const base =
    args[2] === undefined
      ? "ghcr.io/kshramt/evidence_based_scheduling"
      : args[2];
  const ignore_cache_error =
    args[3] === undefined ? false : !!JSON.parse(args[3]);

  run(sha, ref_b64, base, ignore_cache_error);
};

main(Deno.args);
